pkg/tso/keyspace_group_manager.go

```diff
@@ -305,218 +337,20 @@ func (kgm *KeyspaceGroupManager) Close() {
+func (kgm *KeyspaceGroupManager) startRetryUpdateLoop() {
 	defer logutil.LogPanic()
 	defer kgm.wg.Done()
-
-	// Repeatedly watch/apply keyspace group membership/distribution changes until the context is canceled.
+	ticker := time.NewTicker(defaultRetryInterval)
+	defer ticker.Stop()
 	for {
 		select {
 		case <-kgm.ctx.Done():
 			return
-		default:
-		}
-
-		nextRevision, err := kgm.watchKeyspaceGroupsMetaChange(revision)
-		if err != nil {
-			log.Error("watcher canceled unexpectedly and a new watcher will start after a while",
-				zap.Int64("next-revision", nextRevision),
-				zap.Time("retry-at", time.Now().Add(watchEtcdChangeRetryInterval)),
-				zap.Error(err))
-			time.Sleep(watchEtcdChangeRetryInterval)
-		}
-	}
-}
-
-// watchKeyspaceGroupsMetaChange watches any change in keyspace group membership/distribution
-// and apply the change dynamically.
-func (kgm *KeyspaceGroupManager) watchKeyspaceGroupsMetaChange(revision int64) (int64, error) {
-	watcher := clientv3.NewWatcher(kgm.etcdClient)
-	defer watcher.Close()
-
-	ksgPrefix := strings.Join([]string{kgm.legacySvcRootPath, endpoint.KeyspaceGroupIDPrefix()}, "/")
-	log.Info("start to watch keyspace group meta change", zap.Int64("revision", revision), zap.String("prefix", ksgPrefix))
-
-	for {
-		watchChan := watcher.Watch(kgm.ctx, ksgPrefix, clientv3.WithPrefix(), clientv3.WithRev(revision))
-		for wresp := range watchChan {
-			if wresp.CompactRevision != 0 {
-				log.Warn("Required revision has been compacted, the watcher will watch again with the compact revision",
-					zap.Int64("required-revision", revision),
-					zap.Int64("compact-revision", wresp.CompactRevision))
-				revision = wresp.CompactRevision
-				break
-			}
-			if wresp.Err() != nil {
-				log.Error("watch is canceled or closed",
-					zap.Int64("required-revision", revision),
-					errs.ZapError(errs.ErrEtcdWatcherCancel, wresp.Err()))
-				return revision, wresp.Err()
-			}
-			for _, event := range wresp.Events {
-				groupID, err := endpoint.ExtractKeyspaceGroupIDFromPath(string(event.Kv.Key))
-				if err != nil {
-					log.Warn("failed to extract keyspace group ID from the key path",
-						zap.String("key-path", string(event.Kv.Key)), zap.Error(err))
-					continue
-				}
-
-				switch event.Type {
-				case clientv3.EventTypePut:
-					group := &endpoint.KeyspaceGroup{}
-					if err := json.Unmarshal(event.Kv.Value, group); err != nil {
-						log.Warn("failed to unmarshal keyspace group",
-							zap.Uint32("keyspace-group-id", groupID),
-							zap.Error(errs.ErrJSONUnmarshal.Wrap(err).FastGenWithCause()))
-						break
-					}
-					kgm.updateKeyspaceGroup(group)
-				case clientv3.EventTypeDelete:
-					kgm.deleteKeyspaceGroup(groupID)
-				}
-			}
-			// Retry the groups that are not initialized successfully before.
+		case <-ticker.C:
 			for id, group := range kgm.groupUpdateRetryList {
 				delete(kgm.groupUpdateRetryList, id)
```

Comment:
kgm.groupUpdateRetryList seems to be accessed by multiple goroutines now, and it needs to be sync.Map or to add lock.

Reply:
Previously, kgm.groupUpdateRetryList is only accessed by one goroutine.